src/Entities/Player.cpp

```diff
@@ -462,7 +462,17 @@ void cPlayer::SetTouchGround(bool a_bTouchGround)
 	static const auto HalfWidth = GetWidth() / 2;
 	static const auto EPS = 0.0001;
 
-	BLOCKTYPE BlockAtFoot = GetWorld()->GetBlock(POS_TOINT);
+	/* Since swimming is decided in a tick and is asynchronous to this, we have to check for dampeners ourselves.
+	The behaviour as of 1.8.8 is the following:
+	- Landing in water alleviates all fall damage
+	- Passing through any liquid (water + lava) and cobwebs "slows" the player down,
+		i.e. resets the fall distance to that block, but only after checking for fall damage
+		(this means that plummeting into lava will still kill the player via fall damage, although cobwebs
+		will slow players down enough to have multiple updates that keep them alive)
+
+	With this in mind, we first check the block at the player's feet, and decide which behaviour we want to go with.
+	*/
+	BLOCKTYPE BlockAtFoot = ((GetPosY() >= 1) && (GetPosY() <= 255)) ? GetWorld()->GetBlock(POS_TOINT) : E_BLOCK_AIR;
```

Comment:
I did a little work on this bit that cleans the if-nado underneath it up slightly (cf. the fall-damage-on-stairs issue) but that'll take a bit of time to get right since it needs a BlockHandler-level change.
Until then, this'll have to do.